File: nexus-control-plane/src/store/lanes.ts
import { create } from "zustand";
import { invoke } from "@tauri-apps/api/core";

export interface Lane {
  id: string;
  name: string;
  emoji: string;
  color: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

interface LanesState {
  lanes: Lane[];
  loading: boolean;
  error: string | null;
  fetchLanes: () => Promise<void>;
  updateLane: (id: string, name: string) => Promise<void>;
  deleteLane: (id: string) => Promise<void>;
  reorderLanes: (orderedIds: string[]) => Promise<void>;
}

export const useLanesStore = create<LanesState>((set) => ({
  lanes: [],
  loading: false,
  error: null,
  fetchLanes: async () => {
    set({ loading: true, error: null });
    try {
      const lanes = await invoke<Lane[]>("list_lanes");
      set({ lanes, loading: false });
    } catch (e) {
      set({ error: String(e), loading: false });
    }
  },
  updateLane: async (id, name) => {
    const updated = await invoke<Lane>("update_lane", { id, name });
    set((s) => ({
      lanes: s.lanes.map((l) => (l.id === updated.id ? updated : l)),
    }));
  },
  deleteLane: async (id) => {
    await invoke("delete_lane", { id });
    set((s) => ({ lanes: s.lanes.filter((l) => l.id !== id) }));
  },
  reorderLanes: async (orderedIds) => {
    const order = orderedIds.map((id, i) => ({
      id,
      sort_order: (i + 1) * 1000,
    }));
    await invoke("reorder_lanes", { order });
    set((s) => {
      const updated = s.lanes.map((l) => {
        const o = order.find((item) => item.id === l.id);
        return o ? { ...l, sort_order: o.sort_order } : l;
      });
      updated.sort((a, b) => a.sort_order - b.sort_order);
      return { lanes: updated };
    });
  },
}));

File: nexus-control-plane/src-tauri/src/commands/lanes.rs
use nexus_core::db::DbState;
use nexus_core::services::lanes as lane_svc;
use nexus_core::types::{Lane, LaneOrder};

#[tauri::command]
pub fn list_lanes(state: tauri::State<'_, DbState>) -> Result<Vec<Lane>, String> {
    lane_svc::list_lanes(&state)
}

#[tauri::command]
pub fn update_lane(state: tauri::State<'_, DbState>, id: String, name: String) -> Result<Lane, String> {
    lane_svc::update_lane(&state, &id, &name)
}

#[tauri::command]
pub fn delete_lane(state: tauri::State<'_, DbState>, id: String) -> Result<(), String> {
    lane_svc::delete_lane(&state, &id)
}

#[tauri::command]
pub fn reorder_lanes(state: tauri::State<'_, DbState>, order: Vec<LaneOrder>) -> Result<(), String> {
    lane_svc::reorder_lanes(&state, &order)
}
